Content.Server/SimpleStation14/BloodstreamFiller/Systems/BloodstreamFillerSystem.cs

```diff
@@ -0,0 +1,263 @@
+        // {
+        //     _popup.PopupCursor(Loc.GetString(fillComp.TargetInvalidPopup, ("filler", filler)), user);
+        //     return;
+        // }
+
+        var bloodComp = EntityManager.GetComponent<BloodstreamComponent>(target);
+
+        if (!_solution.TryGetSolution(filler, fillComp.Solution!, out var fillerSolution)) // No solution
+            return;
+
+        if (fillComp.Reagent != null && fillComp.Reagent != bloodComp.BloodReagent) // Wrong reagent as specified by the component
+        {
+            _popup.PopupCursor(Loc.GetString(fillComp.TargetInvalidPopup, ("filler", filler)), user);
+            return;
+        }
+
+        if (fillerSolution.Contents.Count == 0) // Empty
+        {
+            _popup.PopupCursor(Loc.GetString(fillComp.EmptyPopup, ("filler", filler)), user);
+            return;
+        }
+
+        if (fillerSolution.Contents.Count > 1) // Extra dorty
+        {
+            _popup.PopupCursor(Loc.GetString(fillComp.DirtyPopup, ("volume", filler)), user);
+            return;
+        }
+
+        if (fillerSolution.Contents[0].ReagentId != bloodComp.BloodReagent) // Wrong reagent contained
+        {
+            _popup.PopupCursor(Loc.GetString(fillComp.TargetBloodInvalidPopup, ("filler", filler), ("target", target)), user);
+            return;
+        }
+
+        var overfill = false;
+
+        var delay = fillComp.FillTime;
+        if (user == target)
+            delay *= fillComp.SelfFillMutli;
+
+        // If the bloodstream is already full, and the filler can overfill, and target is not the user, then overfill.
+        if (fillComp.Overfill && bloodComp.BloodSolution.AvailableVolume == 0 && user != target)
+        {
+            overfill = true;
+            delay *= fillComp.OverfillMutli;
+        }
+
+        _doAfter.TryStartDoAfter(new DoAfterArgs(user, delay, new BloodstreamFillerDoAfterEvent(overfill), target, target, filler)
+        {
+            BreakOnDamage = true,
+            BreakOnTargetMove = true,
+            BreakOnUserMove = true,
+            MovementThreshold = 0.2f,
+            CancelDuplicate = true
+        });
+
+        if (!overfill)
+        {
+            _popup.PopupCursor(Loc.GetString(fillComp.UsePopup, ("target", target)), user);
+            _popup.PopupEntity(Loc.GetString(fillComp.UsedOnPopup), target, target, PopupType.Medium);
+        }
+        else
+        {
+            _popup.PopupCursor(Loc.GetString(fillComp.TargetOverfillPopup, ("target", target)), user, PopupType.MediumCaution);
+            _popup.PopupEntity(Loc.GetString(fillComp.OverfilledPopup), target, target, PopupType.LargeCaution);
+        }
+    }
+
+    private void OnDoAfter(EntityUid uid, BloodstreamComponent component, BloodstreamFillerDoAfterEvent args)
+    {
+        if (args.Cancelled)
+        {
+            return;
+        }
+
+        if (args.Handled || args.Args.Target == null)
+            return;
```

Comment:
can be combined to one
```suggestion
        if (args.Cancelled ||
            args.Handled ||
            args.Args.Target == null)
            return;
```

Reply:
For readability, and separation, I usually see these kept as their own thing at the top. That said, cancelled and handled can probably be combined.